File: download/src/main/java/com/bbou/download/MD5Downloader.java
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */

package com.bbou.download;

import android.util.Log;

import com.bbou.concurrency.Task;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * MD5 downloader task
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
class MD5Downloader extends Task<String, Void, String>
{
	static private final String TAG = "MD5Downloader";

	/**
	 * ResultListener
	 */
	private final Listener listener;

	/**
	 * Exception while executing
	 */
	@Nullable
	private Exception exception;

	/**
	 * Constructor
	 *
	 * @param listener listener
	 */
	MD5Downloader(final Listener listener)
	{
		this.listener = listener;
		this.exception = null;
	}

	@Nullable
	@Override
	protected String doInBackground(final String... params)
	{
		final String md5Arg = params[0];
		final String targetArg = params[1];
		HttpURLConnection httpConnection = null;
		try
		{
			// connect
			final URL url = new URL(md5Arg);
			Log.d(TAG, "Getting " + url);
			final URLConnection connection = url.openConnection();
			connection.connect();

			// expect HTTP 200 OK, so we don't mistakenly save error report instead of the file
			if (connection instanceof HttpURLConnection)
			{
				httpConnection = (HttpURLConnection) connection;
				if (httpConnection.getResponseCode() != HttpURLConnection.HTTP_OK)
				{
					final String message = "server returned HTTP " + httpConnection.getResponseCode() + " " + httpConnection.getResponseMessage();
					throw new RuntimeException(message);
				}
			}

			// open the reader
			try (InputStream is = connection.getInputStream(); InputStreamReader isr = new InputStreamReader(is); BufferedReader reader = new BufferedReader(isr))
			{
				// read
				String line;
				while ((line = reader.readLine()) != null)
				{
					if (line.contains(targetArg))
					{
						final String[] fields = line.split("\\s+");
						return fields[0].trim();
					}
					// cooperative exit
					if (isCancelled())
					{
						Log.d(TAG, "Cancelled!");
						throw new InterruptedException("cancelled");
					}
					if (Thread.interrupted())
					{
						Log.d(TAG, "Interrupted!");
						final InterruptedException ie = new InterruptedException("interrupted while downloading");
						this.exception = ie;
						throw ie;
					}
				}
			}
			return null;
		}
		catch (@NonNull final InterruptedException e)
		{
			this.exception = e;
			Log.d(TAG, e.toString());
		}
		catch (@NonNull final Exception e)
		{
			this.exception = e;
			Log.e(TAG, "While downloading", e);
		}
		finally
		{
			if (httpConnection != null)
			{
				httpConnection.disconnect();
			}
		}
		return null;
	}

	@Override
	protected void onPostExecute(final String result)
	{
		Log.d(TAG, "Completed " + result);
		this.listener.onDone(result);
	}

	@Override
	protected void onCancelled(final String result)
	{
		Log.d(TAG, "Cancelled " + result);
		this.listener.onDone(result);
	}

	/**
	 * Get Exception
	 *
	 * @return exception
	 */
	@Nullable
	synchronized public Exception getException()
	{
		final Exception result = this.exception;
		this.exception = null;
		return result;
	}

	/**
	 * MD5Downloader listener
	 */
	@FunctionalInterface
	public interface Listener
	{
		/**
		 * Done
		 *
		 * @param result md5 digest
		 */
		void onDone(final String result);
	}
}

File: download/src/main/java/com/bbou/download/workers/utils/MD5Downloader.kt
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>.
 */
package com.bbou.download.workers.utils

import android.util.Log
import com.bbou.concurrency.Task
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * MD5 downloader task
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
internal class MD5Downloader(private val listener: Listener) : Task<Pair<String, String>, Void, String?>() {

    /**
     * Exception while executing
     */
    private var exception: Exception? = null

    override fun doJob(params: Pair<String, String>?): String? {
        val md5Arg = params!!.first
        val targetArg = params.second
        var httpConnection: HttpURLConnection? = null
        try {
            // connect
            val url = URL(md5Arg)
            Log.d(TAG, "Getting $url")
            val connection = url.openConnection()
            connection.connect()

            // expect HTTP 200 OK, so we don't mistakenly save error report instead of the file
            if (connection is HttpURLConnection) {
                httpConnection = connection
                if (httpConnection.responseCode != HttpURLConnection.HTTP_OK) {
                    val message = "server returned HTTP " + httpConnection.responseCode + " " + httpConnection.responseMessage
                    throw RuntimeException(message)
                }
            }
            connection.getInputStream().use { `is` ->
                InputStreamReader(`is`).use { isr ->
                    BufferedReader(isr).use { reader ->
                        reader.lineSequence().forEach { it ->
                            if (it.contains(targetArg)) {
                                val fields = it.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                                return fields[0].trim { it <= ' ' }
                            }
                            // cooperative exit
                            if (isCancelled()) {
                                Log.d(TAG, "Cancelled!")
                                throw InterruptedException("cancelled")
                            }
                            if (Thread.interrupted()) {
                                Log.d(TAG, "Interrupted!")
                                val ie = InterruptedException("interrupted while downloading")
                                exception = ie
                                throw ie
                            }
                        }
                    }
                }
            }
            return null
        } catch (e: InterruptedException) {
            exception = e
            Log.d(TAG, e.toString())
        } catch (e: Exception) {
            exception = e
            Log.e(TAG, "While downloading", e)
        } finally {
            httpConnection?.disconnect()
        }
        return null
    }

    override fun onDone(result: String?) {
        Log.d(TAG, "Completed $result")
        listener.onDone(result)
    }

    override fun onCancelled() {
        Log.d(TAG, "Cancelled")
        listener.onDone("???")
    }

    /**
     * MD5 downloader listener
     */
    fun interface Listener {
        /**
         * Done
         *
         * @param result md5 digest
         */
        fun onDone(result: String?)
    }

    companion object {
        private const val TAG = "MD5Downloader"
    }
}
